aptos-move/framework/aptos-framework/sources/multisig_account.move

```diff
@@ -0,0 +1,416 @@
+        assert!(vector::length(&args_hash) > 0, error::invalid_argument(EARGUMENTS_HASH_CANNOT_BE_EMPTY));
+
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        assert_is_owner(owner, multisig_account_resource);
+
+        let transaction = MultisigTransaction {
+            payload: option::none<TransactionPayload>(),
+            payload_hash: option::some(PayloadHash { function_hash, args_hash }),
+            num_approvals: 1,
+            creator: address_of(owner),
+        };
+        add_transaction(multisig_account_resource, transaction);
+    }
+
+    /// Approve a multisig transaction
+    public entry fun approve_transaction(
+        owner: &signer, multisig_account: address, transaction_id: u64) acquires MultisigAccount {
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        assert_is_owner(owner, multisig_account_resource);
+
+        assert!(
+            table::contains(&multisig_account_resource.transactions, transaction_id),
+            error::not_found(ETRANSACTION_NOT_FOUND),
+        );
+        let transaction = table::borrow_mut(&mut multisig_account_resource.transactions, transaction_id);
+        transaction.num_approvals = transaction.num_approvals + 1;
+        emit_event(
+            &mut multisig_account_resource.approve_transaction_events,
+            ApproveTransactionEvent {
+                transaction_id,
+                num_signatures: transaction.num_approvals,
+            }
+        );
+    }
+
+    /// Execute a transaction. This doesn't actually invoke the target function but simply marks the transaction as
+    /// already been executed (by removing it from the transactions table). Actual function invocation is done as part
+    /// executing the MultisigTransaction.
+    ///
+    /// This function is private so no other code can call this beside the VM itself as part of MultisigTransaction.
+    ///
+    /// @param transaction_id Id of the transaction to execute.
+    /// @param target_function Optional and can be empty if the full transaction payload is stored on chain.
+    /// @param args Optional and can be empty if the full transaction payload is stored on chain.
+    /// @return The transaction payload to execute as the multisig account.
+    fun execute_transaction(
+        multisig_account: address,
+        transaction_id: u64,
+        target_function: String,
+        args: vector<u8>,
+    ): TransactionPayload acquires MultisigAccount {
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        assert!(
+            table::contains(&multisig_account_resource.transactions, transaction_id),
```

Comment:
Should the transaction to be executed in order?

Reply:
This is a functionality choice. We can go either way. What do you think?

Reply:
I would personally recommend to focus more on Security than Flexibility in the multi-sig use case. Also some transactions need to be executed in order, and the tx_id offers a native sequence number to do so. 

It's almost the same as the single signed wallet - it does not provide this kind of flexibility either.

Reply:
I'll think more about it. For now, in-order execution is fine.